File: src/utils/tableauParser.ts
export interface LayoutObject {
  id: string;
  type: string;
  name?: string;
  children: LayoutObject[];
  padding?: { top: number; bottom: number; left: number; right: number };
  size?: { width: number; height: number };
  position?: { x: number; y: number };
  attributes?: Record<string, string>;
}

export function parseTwb(xmlString: string): LayoutObject | null {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(xmlString, 'text/xml');

  // Tableau dashboards are typically inside <dashboard> tags
  const dashboard = xmlDoc.querySelector('dashboard');
  if (!dashboard) return null;

  // Extract dashboard size
  const sizeElement = dashboard.querySelector('size');
  let dashboardSize = undefined;
  if (sizeElement) {
    const w = sizeElement.getAttribute('maxwidth') || sizeElement.getAttribute('minwidth');
    const h = sizeElement.getAttribute('maxheight') || sizeElement.getAttribute('minheight');
    if (w && h) {
      dashboardSize = { width: parseInt(w, 10), height: parseInt(h, 10) };
    }
  }

  const rootZone = dashboard.querySelector('zone');
  if (!rootZone) return null;

  const rootLayout = parseZone(rootZone);
  
  // Override root size if dashboard size found
  if (dashboardSize) {
    rootLayout.size = dashboardSize;
  }
  
  return rootLayout;
}

function parseZone(zone: Element): LayoutObject {
  const attributes: Record<string, string> = {};
  for (let i = 0; i < zone.attributes.length; i++) {
    const attr = zone.attributes[i];
    attributes[attr.name] = attr.value;
  }

  const type = zone.getAttribute('type') || zone.getAttribute('param') || 'unknown';
  const id = zone.getAttribute('id') || 'unknown';
  const name = zone.getAttribute('name') || undefined;

  const layoutObject: LayoutObject = {
    id,
    type,
    name,
    children: [],
    attributes,
  };

  const w = zone.getAttribute('w');
  const h = zone.getAttribute('h');
  if (w && h) {
    layoutObject.size = { width: parseInt(w, 10), height: parseInt(h, 10) };
  }

  const x = zone.getAttribute('x');
  const y = zone.getAttribute('y');
  if (x && y) {
    layoutObject.position = { x: parseInt(x, 10), y: parseInt(y, 10) };
  }

  // Try to parse size/padding from 'style' or 'formatted-text' or similar attributes if present
  const style = zone.getAttribute('style');
  if (style) {
    const paddingMatch = style.match(/padding:\s*(\d+)px/);
    if (paddingMatch) {
      const p = parseInt(paddingMatch[1], 10);
      layoutObject.padding = { top: p, bottom: p, left: p, right: p };
    }
  }

  // Parse children zones
  const childZones = zone.querySelectorAll(':scope > zone');
  childZones.forEach((child) => {
    layoutObject.children.push(parseZone(child));
  });

  return layoutObject;
}

File: src/utils/tableau_parser.py
import xml.etree.ElementTree as ET
import json

def parse_tableau_twb(xml_content):
    root = ET.fromstring(xml_content)
    
    # 1. Locate the first dashboard
    dashboards = root.findall('.//dashboards/dashboard')
    if not dashboards:
        return "No dashboard found."
    
    dashboard = dashboards[0]
    db_name = dashboard.get('name', 'Unknown')
    
    # 2. Get total size
    size_el = dashboard.find('size')
    width = int(size_el.get('maxwidth', 1900))
    height = int(size_el.get('maxheight', 1400))
    
    # 3. Parse zones
    zones_data = []
    
    def parse_zone(zone, parent_id=None):
        zone_id = zone.get('id')
        name = zone.get('name')
        
        # Determine type
        param = zone.get('param')
        zone_type = param if param else ('worksheet' if zone.find('zone') is None else 'container')
        
        # Geometry conversion
        x = (int(zone.get('x', 0)) / 100000) * width
        y = (int(zone.get('y', 0)) / 100000) * height
        w = (int(zone.get('w', 0)) / 100000) * width
        h = (int(zone.get('h', 0)) / 100000) * height
        
        zone_info = {
            "ID": zone_id,
            "Name": name,
            "Type": zone_type,
            "Geometry": {"X": x, "Y": y, "W": w, "H": h},
            "Fixed-Size": zone.get('is-fixed') == 'true',
            "Hierarchy": parent_id
        }
        zones_data.append(zone_info)
        
        for child in zone.findall('zone'):
            parse_zone(child, zone_id)

    root_zone = dashboard.find('zone')
    if root_zone is not None:
        parse_zone(root_zone)
        
    return {"Dashboard": db_name, "Size": {"W": width, "H": height}, "Zones": zones_data}

# Example usage (assuming xml_content is loaded):
# result = parse_tableau_twb(xml_content)
# print(json.dumps(result, indent=2))
